serviceregistration/kubernetes/client/client.go

```diff
@@ -0,0 +1,288 @@
+		shouldRetry, err := c.attemptRequest(client, req, ptrToReturnObj)
+		if !shouldRetry {
+			// The error may be nil or populated depending on whether the
+			// request was successful.
+			return err
+		}
+		lastErr = err
+	}
+	return lastErr
+}
+
+// attemptRequest tries one single request. It's in its own function so each
+// response body can be closed before returning, which would read awkwardly if
+// executed in a loop.
+func (c *Client) attemptRequest(client *http.Client, req *http.Request, ptrToReturnObj interface{}) (shouldRetry bool, err error) {
+	// Preserve the original request body so it can be viewed for debugging if needed.
+	// Reading it empties it, so we need to re-add it afterwards.
+	var reqBody []byte
+	if req.Body != nil {
+		reqBody, _ = ioutil.ReadAll(req.Body)
+		reqBodyReader := bytes.NewReader(reqBody)
+		req.Body = ioutil.NopCloser(reqBodyReader)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return false, err
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			if c.logger.IsWarn() {
+				// Failing to close response bodies can present as a memory leak so it's
+				// important to surface it.
+				c.logger.Warn(fmt.Sprintf("unable to close response body: %s", err))
+			}
+		}
+	}()
+
+	// Check for success.
+	switch resp.StatusCode {
+	case 200, 201, 202:
+		// Pass.
+	case 401, 403:
+		// Perhaps the token from our bearer token file has been refreshed.
+		config, err := inClusterConfig()
+		if err != nil {
+			return false, err
+		}
+		if config.BearerToken == c.config.BearerToken {
+			// It's the same token.
+			return false, fmt.Errorf("bad status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+		}
+		c.config = config
+		// Continue to try again, but return the error too in case the caller would rather read it out.
+		return true, fmt.Errorf("bad status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+	case 404:
+		return false, &ErrNotFound{debuggingInfo: sanitizedDebuggingInfo(req, reqBody, resp)}
+	case 500, 502, 503, 504:
+		// Could be transient.
+		return true, fmt.Errorf("unexpected status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+	default:
+		// Unexpected.
+		return false, fmt.Errorf("unexpected status code: %s", sanitizedDebuggingInfo(req, reqBody, resp))
+	}
+
+	// We only arrive here with success.
+	// If we're not supposed to read out the body, we have nothing further
+	// to do here.
+	if ptrToReturnObj == nil {
+		return false, nil
+	}
+
+	// Attempt to read out the body into the given return object.
+	return false, json.NewDecoder(resp.Body).Decode(ptrToReturnObj)
+}
+
+type Pod struct {
+	Metadata *Metadata `json:"metadata,omitempty"`
+}
+
+type Metadata struct {
+	Name string `json:"name,omitempty"`
+
+	// This map will be nil if no "labels" key was provided.
+	// It will be populated but have a length of zero if the
+	// key was provided, but no values.
+	Labels map[string]string `json:"labels,omitempty"`
+}
+
+type PatchOperation int
```

Comment:
Can this be `type PatchOperation string` instead? The constants can simply be Add = "add", ..., and then the Parse and String operations aren't needed.

Reply:
Hm, last I checked it didn't actually work that way in Go, but I will check again in case anything has changed.

Reply:
Actually, that turned out really nicely: https://github.com/hashicorp/vault/pull/8249/commits/74c438ebaad33111ceb5f6e08f34856b529dfab5